bindings/rust/s2n-tls-tokio/tests/tcp.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use s2n_tls::connection::Builder;
+use s2n_tls_tokio::{TlsAcceptor, TlsConnector};
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+
+pub mod common;
+
+trait TestStream: AsyncRead + AsyncWrite + Unpin {}
+impl<T: AsyncRead + AsyncWrite + Unpin> TestStream for T {}
+
+type TestPair = (Box<dyn TestStream>, Box<dyn TestStream>);
+type TestPairList = Vec<TestPair>;
+async fn new_test_pairs<A, B>(
+    server_builder: A,
+    client_builder: B,
+) -> Result<TestPairList, Box<dyn std::error::Error>>
+where
```

Comment:
I'm pretty attached to the idea of an array of things that should behave the same way though. If we have different test cases for TCP streams vs TLS streams, then they could end up testing for different behaviors / allowing differences. There isn't an enforcement mechanism that the behavior is the same, which is the point of this test suite.

Like, I'd love us to expand these tests to cover checking that behaviors other than handling EOF are the same between the streams.

I wonder if I could add TLS1.2 as a third element in the array tho 🤔

Reply:
And while the implementation is pretty complex, calling it is pretty simple, so I'm not super worried about the complexity.

Reply:
> If we have different test cases for TCP streams vs TLS streams, then they could end up testing for different behaviors / allowing differences.

My thinking was that you could enforce that through an assert method. 
```
assert_write_to_closed_behavior<S>(reader: S, writer: S) 
where
    S: AsyncRead + AsyncWrite // we can use a generic here to avoid relying on TLS specifics
{
    drop(writer);
    let result = reader.read_u8().await;
    assert!(result.is_err());
    let error = result.unwrap_err();
    assert!(error.kind() == std::io::ErrorKind::UnexpectedEof);
}
```

And then your actual unit test could be something like
```
read_from_closed_behavior() {
    let (client_tcp, server_tcp) = get_streams();
    assert_write_to_closed_behavior(client_tcp, server_tcp)
    let (client_tls13, server_tls13) = ...
    assert_write_to_closed_behavior(client_tls13, server_tls13)
}
```

Reply:
I made this change, but idk if I like it. It moves the unique logic (like assert_write_to_closed_behavior) into a separate method while every test has to repeat the common setup logic (creating streams). It was only really reasonable when I created another helper method to abstract away tls stream creation.

Reply:
And if I want to add another kind of stream, I'm going to have to modify every test rather than just the stream setup method.

Reply:
I like that the checks are now abstracted out into their own little unit of logic, but I see your concern about the stream setup. I think we can hurdle that when we come to it. There is a round-about way of defining a stream harness that takes in the "behavior assertion" closure that would get rid of the duplication, but if the streams keep increasing I agree that the dyn list approach might make more sense.